File: bindings/java/src/main/java/io/pap/PrincipalKeypair.java
package io.pap;

import com.sun.jna.Pointer;

/**
 * Root Ed25519 keypair bound to the human principal.
 * Implements {@link AutoCloseable} for use in try-with-resources.
 */
public final class PrincipalKeypair implements AutoCloseable {

    private final Pointer handle;

    private PrincipalKeypair(Pointer h) {
        this.handle = h;
    }

    /** Generate a new random principal keypair. */
    public static PrincipalKeypair generate() {
        Pointer h = PapLib.INSTANCE.pap_keypair_generate();
        if (h == null) throw PapException.fromLastError("keypair_generate");
        return new PrincipalKeypair(h);
    }

    /**
     * Reconstruct from 32 raw secret-key bytes.
     *
     * @param bytes exactly 32 bytes
     */
    public static PrincipalKeypair fromSecretBytes(byte[] bytes) {
        if (bytes.length != 32)
            throw new IllegalArgumentException("Secret key must be exactly 32 bytes");
        Pointer h = PapLib.INSTANCE.pap_keypair_from_bytes(bytes, bytes.length);
        if (h == null) throw PapException.fromLastError("keypair_from_bytes");
        return new PrincipalKeypair(h);
    }

    /** The {@code did:key} identifier derived from this keypair. */
    public String did() {
        Pointer ptr = PapLib.INSTANCE.pap_keypair_did(handle);
        if (ptr == null) throw PapException.fromLastError("keypair_did");
        String s = ptr.getString(0);
        PapLib.INSTANCE.pap_string_free(ptr);
        return s;
    }

    /** The raw 32-byte public key. */
    public byte[] publicKeyBytes() {
        byte[] buf = new byte[32];
        if (PapLib.INSTANCE.pap_keypair_public_bytes(handle, buf) != 0)
            throw PapException.fromLastError("keypair_public_bytes");
        return buf;
    }

    /** Sign arbitrary bytes; returns the 64-byte Ed25519 signature. */
    public byte[] sign(byte[] message) {
        byte[] sig = new byte[64];
        if (PapLib.INSTANCE.pap_keypair_sign(handle, message, message.length, sig) != 0)
            throw PapException.fromLastError("keypair_sign");
        return sig;
    }

    /** Package-private: expose the raw Pointer for use by other classes. */
    Pointer raw() { return handle; }

    @Override
    public void close() {
        PapLib.INSTANCE.pap_keypair_free(handle);
    }
}

File: bindings/csharp/PrincipalKeypair.cs
using System.Runtime.InteropServices;
using Microsoft.Win32.SafeHandles;

namespace Pap;

/// <summary>
/// Root Ed25519 keypair bound to the human principal.
/// Implements <see cref="IDisposable"/>; wraps the native handle in a
/// <see cref="SafeHandle"/> for deterministic cleanup.
/// </summary>
public sealed class PrincipalKeypair : IDisposable
{
    private sealed class Handle : SafeHandleZeroOrMinusOneIsInvalid
    {
        internal Handle(IntPtr h) : base(ownsHandle: true) { SetHandle(h); }
        protected override bool ReleaseHandle() { PapNative.pap_keypair_free(handle); return true; }
    }

    private readonly Handle _handle;

    internal IntPtr Raw => _handle.DangerousGetHandle();

    private PrincipalKeypair(IntPtr h) => _handle = new Handle(h);

    /// <summary>Generate a new random principal keypair.</summary>
    public static PrincipalKeypair Generate()
    {
        var h = PapNative.pap_keypair_generate();
        if (h == IntPtr.Zero) PapNative.ThrowLastError("keypair_generate");
        return new PrincipalKeypair(h);
    }

    /// <summary>Reconstruct from 32 raw secret-key bytes.</summary>
    public static PrincipalKeypair FromSecretBytes(byte[] bytes)
    {
        if (bytes.Length != 32)
            throw new ArgumentException("Secret key must be exactly 32 bytes.", nameof(bytes));
        var h = PapNative.pap_keypair_from_bytes(bytes, (nuint)bytes.Length);
        if (h == IntPtr.Zero) PapNative.ThrowLastError("keypair_from_bytes");
        return new PrincipalKeypair(h);
    }

    /// <summary>The <c>did:key</c> identifier derived from this keypair.</summary>
    public string Did() => PapNative.OwnedString(PapNative.pap_keypair_did(Raw));

    /// <summary>The raw 32-byte public key.</summary>
    public byte[] PublicKeyBytes()
    {
        var buf = new byte[32];
        if (PapNative.pap_keypair_public_bytes(Raw, buf) != 0)
            PapNative.ThrowLastError("keypair_public_bytes");
        return buf;
    }

    /// <summary>Sign arbitrary bytes; returns the 64-byte Ed25519 signature.</summary>
    public byte[] Sign(byte[] message)
    {
        var sig = new byte[64];
        if (PapNative.pap_keypair_sign(Raw, message, (nuint)message.Length, sig) != 0)
            PapNative.ThrowLastError("keypair_sign");
        return sig;
    }

    public void Dispose() => _handle.Dispose();
}
